### DISCUSSION TITLE
Coding ligatures

### DISCUSSION THREAD

**Discussion by astrolemonade** [Mark2Mark/gintronic#13] [General]
Lots of programming fonts support ligatures, something that wasn't a thing when Gintronic started. Today they improve readability a lot and sometimes help at saving screen width.
![image](https://user-images.githubusercontent.com/35623921/87952987-22cc6a00-cab3-11ea-9c1c-98b3200ae97e.png)
(image took from [FiraCode repo](https://github.com/tonsky/FiraCode/) ) There are more to be discussed but these are the minimum minimorum.
The above picture shows some important ligatures that one programming font should support for better coding experience.
As a workaround, you can add ligatures to the font from FiraCode(using https://github.com/ToxicFrog/Ligaturizer) which transfers some of the ligatures to Gintronic. The big issue is that they feel a little bit out of place and blurry sometimes!

**Comment by Mark2Mark** [comment]
Thanks for the links and input. I will have a look.

@cata0309 do you also happen to have some dummy code snippets which use as many of those ligatures as possible? When I design them, I need to see them in proper context. The overview from above is super helpful, but the context would be top notch.

Also which code editors do you use? Gintronic was mainly targeted into Sublime Text, Atom, Xcode and Mac Terminal back then. Currently I also use VS occasionally.

**Comment by Mark2Mark** [comment]
The multiply sign in the Hexadecimal Ex is weird: it’s not a `multiply` but should be an `x` in `0xFF`

**Comment by Mark2Mark** [comment]
The equal signs are very debatable. In some programming languages it might be preferred as you see in the `Ligatures On` examples, but many programmers want to see for instance `!=` or `==` as exactly those 2 characters they typed.

**Comment by astrolemonade** [comment]
![image](https://user-images.githubusercontent.com/35623921/88311231-4c86ca80-cd19-11ea-8461-b527be114f92.png)
![image](https://user-images.githubusercontent.com/35623921/88311255-53154200-cd19-11ea-8cc1-3840865a6b36.png)

**Comment by Mark2Mark** [comment]
Can you elaborate and not just post screenshots please? I don’t know what those are about.

Also for the code snippets: I’d need them as text, not as images. Thank you!

**Comment by astrolemonade** [comment]
I can't know all the ligatures use case but you can try testing programming fonts with ligatures on this website:
https://www.programmingfonts.org/ Choose somethink like JetBrainsMono/Iosevka etc

**Comment by mariojackson** [comment]
Hi @Mark2Mark 

I'm thinking about buying the font. I like its uniqueness and style. 

However, the font has its price and before buying it, I'd like to know if ligatures are planned. I would buy it in an instant if ligatures were being added in the future. Otherwise, I would have a hard time, since many other programming fonts which are cheaper, have ligatures.

Can you tell me whether ligatures are planned?

Thanks in advance!

**Comment by Mark2Mark** [comment]
Hi @mariojackson 

Thanks for your request.
~~During latest developments it seems that I wont be likely to add new features to Gintronic as it is. Only bugfixing. However I am working on the predecessor of Gintronic which I will release with my own foundry [Hypertype](https://hypertype.xyz) soon. I am also considering to add ligatures for that one, even though I am a great disadvocat of coding ligatures myself. But I see there are user’s who wants them, so I’ll probably go the extra mile and add them there. I really respect your and other’s request, but also want to mke sure that you don’t buy anything in the hope to get features that does not yet claim to have. When the predecessor has ligatures, I’ll also advertise those as being included.~~
**Update:** I am currently working on version 2.0 of Gintronic, which will include a set of coding ligatures. The current development can be followed here.

I’d be curious:
- Which editors and on which platfrom(s) are you programming?
- Is it okay if ligatures are optional, are those toggle-able in your IDEs?
- Are there more important ligatures than others? Or is is a all or nothing?
- Are there certain levels of ligatures? E.g. the `==` should not be connected in my eyes, it’s changing the semantics too much. However, other ligatures are very plausible, so maybe there should be a way to to go gradually from `no ligas` over `some ligas`all the way to `everything-ligas`?

**Comment by mariojackson** [comment]
Hi @Mark2Mark 

Thank you for your fast reply. I totally understand that ligatures are not for everybody and that it's more subjective. Thank you for being honest, I'll wait for your predecessor in this case. Do you already know around which time it will come out?

- I'm mostly using JetBrains IDE's, VS Code, and Vim on macOS and Linux.
- It is totally ok of course, if ligatures are optional. I would even prefer it.
- I don't have to have all ligatures personally. I'm totally ok if `==` is not connected. I like simple and decent ligatures, in cases where they'd reduce clutter like having 
<img width="139" alt="image" src="https://user-images.githubusercontent.com/57389555/136531795-24517124-def0-4bdf-9180-f2ba0a82316f.png"> instead of `--> -> -->>`
- I think it's a good idea. I don't like all ligatures, too. Especially not if they make things harder to read (like `==`, `===` or `www`).  So something gradually would be great, to satisfy most users imo.

**Comment by Mark2Mark** [comment]
Hi @mariojackson,

Thanks for you input, that’s very helpful.
I hope the release can be early ’22. We’re working with pressure on it.

I guess it would be reasonable to make some sort of list of ligatures and then mark those that are relevant and kick out others that are too much. I don’t even have a proper supervision of which ligatures can/should exist, as I am not working with too many programming languages or syntaxes. For me it is: Objective-C, C, Swift, Python, Java, JavaScript, PHP, CSS, HTML, markdown, shell and so on, but no Haskell, Closure, or Ruby or Go etc, where ligas might come in handy more, I guess.

**Comment by effinsky** [comment]
Is there a place catch a glimpse of the new monospace font?

**Comment by Mark2Mark** [comment]
Not at this very moment. But I hope soon.

**Comment by astrolemonade** [comment]
Looking forward to seeing the progress on the new monospace font too.

**Comment by Mark2Mark** [comment]
Thanks, Cons-Cat. I know what stylistic sets are ;)

BTW: `onum`, `zero` etc are not stylistic sets, but also separate OpenType-Features. Gintronic has similar things for the numbers, just that it is a lining number feature to turn on as the onums are default in this design.

My quoted question was not about the technical implementations of how those ligatures are made (I am well aware of that), but the differentiation of which ligatures should be forced and which should be added in a granular fashion by the user optionally.

**Comment by kkirsche** [comment]
Apologies if this feedback isn't useful, but wanted to throw my feedback in should it be of value.

> * Which editors and on which platfrom(s) are you programming?

I use VS Code primarily with Neovim as a fallback.

> * Is it okay if ligatures are optional, are those toggle-able in your IDEs?

Yes, they are. I learned how to do so via a different font, MonoLisa. See https://www.monolisa.dev/faq#how-to-enable-alternate-zero-stylistic-sets-and-the-script-variant for how they solve this problem (from the editor side, not the font itself).

> * Are there more important ligatures than others? Or is is a all or nothing?

In my opinion, the most important ligatures are ones that address visual kerning, such as `++` and `??`, and ones that are meant to act as "shapes" such as `->` and `=>`. It's not all or nothing for me, and as long as they cover the most common use cases, I don't think it's a big deal.

> * Are there certain levels of ligatures? E.g. the `==` should not be connected in my eyes, it’s changing the semantics too much. However, other ligatures are very plausible, so maybe there should be a way to to go gradually from `no ligas` over `some ligas`all the way to `everything-ligas`?

This is achievable, but I don't know technically how it's implemented by other fonts. I view it from a category mindset, such as `equality` could enable `!=`, `==`, `>=`, etc., `arrows` may enable `->`, `=>`, `<=`, `<-`, etc.

Thanks for all of your time and work. I greatly appreciate all you have done for Gintronic and love it as my terminal typeface.

**Comment by Mark2Mark** [comment]
Thank you, @kkirsche 

Regarding the font features: Funny, I [just added this today](https://github.com/Mark2Mark/gintronic/discussions/11) 

The different levels could be implemented via different stylistic sets. So maybe `ss01` only enables `==`, `??`, `→`, `=>` and `ss02` enables more "advanced/unusal" ligatures, etc.

Many thanks for all your input. I will take it all into account.
Glad that you like Gintronic. I will soon send out an update with some minor fixes as well as the narrower spacing as the default now (some IDE’s or terminals ignored that, so some users didn’t even see it. But in other IDE’s it turned out to look too much spaced appart.

**Comment by effinsky** [comment]
speaking of... there's narrower spacing? please link to how one enables that and/or to some examples of what that looks like. thanks!

**Comment by Mark2Mark** [comment]
It is coming very soon. I’ll send it out once it’s ready. It will be new font files to be replaced with the former ones.

**Comment by Mark2Mark** [comment]
A little note about my current planning/thinking:

In recent times, IDEs made it more feasible to turn on certain OpenType features (see a [growing collection of how to here](https://github.com/Mark2Mark/gintronic/discussions/11)), which usually allow for a more granular control over the alternatives selection.

So I will be trying to direct different sets of ligatures into stylistic sets, where the default (all sets off) will be no ligatures at all; set 1 will have some very common ligas; set 2 will have those, plus more less common ones; and so on. I still need to figure out which will go where.

Another new feature I will implement is contextual GPOS, which allows to move glyphs around in a certain context. This will allow for improved readability without ligatures. For example in `foo::bar` the two colons can move closer together to appear more as a union.

**Comment by zygous** [comment]
I've been using Cartograph for quite a while now (although I've just switched to Gintronic), and quickly came to really like the connected `==`, the slashed `!=` and the triple-bar connected versions of `===` and `!==` (which are heavily used in JavaScript and TypeScript). In practice, there's no confusion between a double-width, connected `==` and a shorter `=` for assignment.

The other things that I've enjoyed are proper arrows, e.g. `->`, `=>`, which improve legibility greatly. Ruby uses `~>` arrows in Gemfiles, and it's nice to see those as proper arrows too (Mono Lisa has this, but Cartograph does not).

The greater-than-or-equal-to operator is nicer to read when rendered as the symbol it's supposed to mimic `>=`. This _does_ lead to a potential conflict for less-than-or-equal-to and left fat arrow: `<=`. In my coding, I've never needed to use a left fat arrow though, so I'd prioritise a ligature for less-than-or-equal-to (I guess this could be overridden with an OpenType feature though).

Other than a those few key ligatures, I think the contextual GPOS would deal with all the other cases I've found are a nice touch, e.g. slightly closer `::`.

TL;DR: The ligatures I'd _most_ like to see are for `->`, `<-`, `~>`, `=>`, `<=`, `>=`, `<=>`, `==`, `===`, `!=`, `!==`. I think that would cover most cases for me using the following languages:

- JavaScript/TypeScript
- PHP
- Ruby
- Rust
- Java
- Bash

(BTW, I mostly use Neovim in Kitty on a Mac. Kitty supports ligatures (obviously) and specifying OpenType font features per-font.)

**Reply by Mark2Mark** [reply]
Thanks for your input!

This looks like a good list to use as a start for the implementation. Seems a nice set for the first level of ligature-ness. 

As I plan to disable ligatures by default, I can just ship updates with these without breaking backwards compatibility. Hoping for you users to help testing and improving. Coding ligatures are hard to test IRL by one person, as they are subjectively to be judged. So I can try that democratic approach 👾

**Reply by zygous** [reply]
Sounds ideal! Thanks Mark 🙌🏻

**Comment by mariojackson** [comment]
Hi @Mark2Mark!  I'm curious about the release date for version 2.0.  Do you happen to know when it's approximately coming out? Also, will there be a free upgrade for those who purchased version 1? I've been waiting patiently, so a little longer isn't a problem, but I would consider buying it now instead, if I don't have to pay for the upgrade of version 2 when it's coming out.

**Reply by Mark2Mark** [reply]
HI @mariojackson. Thanks for your comment.

### About version 2.0:
First of all, yes, the update will be free for anyone who purchased via my website (and has the invoice from Paddle which handles the payment process). If you purchase now, you will definitely eligible to get this update for free.

### About the ETA:
Due to the parental leave I am in since last summer, it is still a little hard to predict, but I strongly plan to get this version out latest by this summer. I don’t want to push too many updates on the users, but I am also considering making smaller increments and maybe one or 2 more updates instead, so that I can get progress to you at least.

Sorry again for the waiting. All your patience is highly appreciated.

**Comment by brzzdev** [comment]
Hi @Mark2Mark,

Just wondering if we could get a status update on how 2.0 is going?

**Reply by Mark2Mark** [reply]
Thanks for pushing. The work on it has just reactivated, and while working on some other releases and client projects I am in the warmup. Hope to send out beta tests soon to anyone interested. 

If you're interested, please drop your names below.

**Reply by zygous** [reply]
Exciting news! I’d like to beta test.

**Reply by brzzdev** [reply]
I’m certainly interested!

**Reply by mariojackson** [reply]
I'm interested as well

**Reply by brzzdev** [reply]
Happy new year! How's it looking @Mark2Mark? 👀

**Reply by Mark2Mark** [reply]
Happy new year everyone and thanks for your patience. 
I am currently working full time on our foundry website and prepare multiple fonts for the release. Gintronic 2.0 is planned to be part of the launch end of March or the first addition after launch. Let’s all cross fingers to get it out soon 💼

**Comment by Mark2Mark** [comment]
Hello everyone following here, I want to share the plan and get your input on a few key decisions.

## Core Principles

- **Default = no ligatures.** Current behaviour is fully preserved. Nothing changes unless you opt in.
- **Graduated levels via stylistic sets.** You pick the categories you want — arrows, equality, comments, etc. — and mix-and-match freely.
- **Subtle spacing improvements always active.** Pairs like `::`, `||`, `&&`, `()` get tighter contextual spacing out of the box, no features to enable. Think of it as optical refinement for code.

---

## Proposed Ligature Categories

Six independent stylistic sets. Enable any combination — they’re not cumulative ”levels,“ they’re categories you freely combine.

### SS01 — Essential Arrows
```
->  <-  =>  ~>  <~
-->  <--  <->  ==>  <==  <=>  ~~>
<-->  <==>
```

### SS02 — Equality & Comparison
```
==  !=  >=  <=
===  !==  =/=
```

### SS03 — Logical & Compound
```
&&  ||  ??  !!  ++  --  **  ::  :::
```

### SS04 — Comment Delimiters
```
//  ///  /*  */  <!--
```

### SS05 — Pipe & Flow
```
|>  <|  <|>  >>=  =<<  <$>  <*>  >=>  :=
```

### SS06 — Extended & Ranges
```
..  ...  ..<  ..=  <<<  >>>  #!  #=
```

### How to enable in your editor

```jsonc
// VS Code
"editor.fontLigatures": "'ss01', 'ss02'"

// Kitty
font_features Gintronic +ss01 +ss02

// CSS
font-feature-settings: "ss01", "ss02";
```

---

## Always-On Spacing Improvements (No SS Required)

These subtle repositioning adjustments make operator pairs read as visual units — tighter `::`, balanced `()`, aligned `==`. Always active, even with all ligatures disabled:

| Pair | What happens |
|------|-------------|
| `()` `{}` `[]` | Inner characters nudge closer |
| `::` | Colons tighten (scope operator) |
| `\|\|` | Pipes tighten (logical OR) |
| `&&` | Ampersands tighten (logical AND) |
| `??` | Question marks tighten (null coalescing) |
| `++` `--` | Signs align as unit |
| `..` | Dots redistribute (but not `...` or `..<`) |
| `!!` | Exclamation marks tighten |
| `:=` `?.` | Tighten as assignment/chaining operators |

---

## The `<=` Problem

This is the most contentious ligature in the industry. `<=` means two different things:

- **≤** (less-than-or-equal) — used in virtually every language
- **⇐** (left fat arrow) — used in Haskell and some markup

My current plan: **default to ≤ (math symbol)**, with an optional character variant (CV01) to switch it to the arrow form.

---

## Open Questions — Your Input Wanted

### 1. Does this category split make sense?

Are there ligatures you’d move between categories? Sequences missing that you use daily? Anything included that you’d never want?

### 2. `<=` — arrow or math symbol?

When you type `<=`, do you expect:
- **(A)** ≤ — less-than-or-equal *(my current default)*
- **(B)** ⇐ — left arrow

### 3. `==` appearance

Should `==` look like:
- **(A)** Two connected bars (classic ligature look)
- **(B)** Two bars with a small gap (lighter, less "fused")
- **(C)** Leave it unchanged (just use spacing adjustment)

### 4. What languages do you code in?

This helps me prioritise. If nobody here writes Haskell, the Haskell-heavy sequences in SS05 can wait. If everyone writes JavaScript, SS01 + SS02 ship first.

### 5. Would you use a “Liga” variant?

Some editors (Xcode, some terminals) can’t enable individual stylistic sets. Would you use a separate font file — "Gintronic Liga" — with all ligatures baked in?

### 6. Other?

Any other ligatures I missed in my collection?

### 7. Infinite arrows — worth it?

Some fonts support arrows of any length (`--->`, `---->`, etc.) via a composable segment system. It’s impressive but adds complexity. Is this something you’d actually use, or are 2–3 character arrows enough?

---

Looking forward to your thoughts. The more concrete the feedback, the better — especially on the open questions above.

**Reply by brzzdev** [reply]
As someone who is not a fan of ligatures and I am in this thread just to see font updates, I would like to be sure that
> Default = no ligatures. Current behaviour is fully preserved. Nothing changes unless you opt in.

means that we can still get a version that has none of these features, or that they are truly opt-in. Working in Xcode, which has horrible OpenType support, it's impossible to disable ligatures and features and I have to reach for 'NL' version like JetBrains Mono and Cascadia Code have.
See https://github.com/0xType/0xProto/issues/116

**Reply by zygous** [reply]
This looks great to me!

> 1. Does this category split make sense?

Looks good. I'd likely use all but SS06, though, so the split doesn't matter much to me.

> 2. <= — arrow or math symbol?

Definitely less-than-or-equal

> 3. == appearance

Two connected bars (classic ligature look)

> 4. What languages do you code in?

JS, TS, PHP, Rust, Python. Not Haskell.

> 5. Would you use a “Liga” variant?

Only if I had to.

> 6. Other?

Nope

> 7. Infinite arrows — worth it?

Infinite === for progress bars, perhaps, but infinite arrows aren't worth it to me.


Thanks for working on this, I'm really looking forward to it.

**Reply by Mark2Mark** [reply]
Yes, its totally opt-in, as stated above. I myself also don’t use coding ligatures at all, this is the entire reason for this thread and why it takes so long, because I can only rely on the input of users who use coding ligatures :)
